File: dotfiles/hammerspoon/.hammerspoon/Spoons/BrowserTabs.spoon/providers/safari.lua
--- === BrowserTabs.providers.safari ===
---
--- The Safari backend. Safari has its own dictionary rather than the Chromium one, so it is
--- its own file, and it owns its bundle identifier the way a provider owns knowledge of its
--- own backend. Three differences from Chromium shape the code. A tab's title is `name`, not
--- `title`. A window names its selected tab through `current tab`, an object, rather than an
--- active tab index, so the index is read back off that object. And a minimized window is
--- `miniaturized` here and `minimized` there.
---
--- Safari tabs carry no id at all, which is the reason tab identity across this whole spoon
--- is the bundle id plus the URL rather than a browser tab id. The consequence and its cost
--- are recorded in the spoon's CLAUDE.md.

local spoonPath = debug.getinfo(1, "S").source:sub(2):match("(.*/)")
local function loadShared(name)
  local chunk, err = loadfile(spoonPath .. "../" .. name)
  if not chunk then
    error("BrowserTabs.providers.safari: failed to load " .. name .. ": " .. tostring(err))
  end
  return chunk()
end

local jxa = loadShared("jxa.lua")
local apps = loadShared("apps.lua")

local BUNDLE_ID = "com.apple.Safari"

-- Every tab of every window, names and URLs fetched in bulk so the cost is one Apple Event
-- per property per window. The running guard keeps a listing from ever launching Safari.
--
-- A window with no document is skipped. Safari keeps at least one window in its dictionary that
-- is not on screen at all, a leftover carrying a single Start Page tab, and it is invisible to
-- the accessibility layer, so a row offered for it can be chosen and then nothing happens, which
-- is one of the ways this tool looked like it was failing. The test is the document rather than
-- the `visible` flag, because a minimized window is not visible either, and neither is any window
-- of an application hidden with Command H, so filtering on that would throw away real tabs. A
-- window genuinely showing the Start Page still has a document, so this only ever drops the
-- phantom, all three cases measured rather than assumed.
local LIST = [==[
function run(argv) {
  const app = Application(argv[0]);
  if (!app.running()) return "[]";
  const out = [];
  const wins = app.windows, n = wins.length;
  for (let w = 0; w < n; w++) {
    const win = wins[w];
    let wid = null, at = -1, titles = [], urls = [];
    try { if (!win.document()) continue; } catch (e) { continue; }
    try { wid = win.id(); } catch (e) {}
    try { at = win.currentTab().index(); } catch (e) {}
    try { titles = win.tabs.name(); urls = win.tabs.url(); } catch (e) { continue; }
    for (let t = 0; t < titles.length; t++) {
      out.push({ windowID: String(wid), windowIndex: w + 1, tabIndex: t + 1,
                 active: (t + 1) === at, title: titles[t], url: urls[t] });
    }
  }
  return JSON.stringify(out);
}
]==]

-- Select a tab and bring its window forward. The window is addressed by id through the shared
-- `windowById`, never by its position, for the reason recorded in jxa.lua, that both of the
-- writes below reorder the window list and a positional specifier would then be pointing at
-- another window. Raising the application is the caller's job.
--
-- A minimized window is restored, since raising the application leaves it in the Dock and the
-- whole thing then looks like it did nothing at all. Safari calls that state `miniaturized`, the
-- Standard Suite spelling, where the Chromium dictionary says `minimized`, which is the third
-- difference that earns this file its own copy of the activation. Ordering between the restore
-- and the reorder was measured both ways on this browser too, and neither is better once the
-- window is addressed by id.
--
-- The window's own name is read back afterwards and returned, since the caller has to find this
-- same window at the accessibility layer. Safari's window ids happen to be the window server's own
-- so the caller usually matches on those, but the name is what covers a window it cannot. Safari
-- names a window after its tab group as well as the selected tab, which is exactly why the tab's
-- title cannot stand in for this.
local ACTIVATE = [==[
function run(argv) {
  const app = Application(argv[0]);
  if (!app.running()) return JSON.stringify({ ok: false });
  const win = windowById(app, argv[1]);
  if (!win) return JSON.stringify({ ok: false });

  // The tab numbers move whenever a tab is opened or closed after the list was built, so the
  // position is checked against the URL and only trusted when the two still agree. The position
  // wins when they do, since a window often holds the same address more than once.
  let idx = parseInt(argv[2], 10);
  try {
    const urls = win.tabs.url();
    const want = argv[3] || "";
    if (want && urls[idx - 1] !== want) {
      const found = urls.indexOf(want);
      if (found >= 0) idx = found + 1;
    }
  } catch (e) {}

  try { win.currentTab = win.tabs[idx - 1]; } catch (e) { return JSON.stringify({ ok: false }); }
  try { if (win.miniaturized()) win.miniaturized = false; } catch (e) {}
  try { win.index = 1; } catch (e) {}
  const r = { ok: true };
  try { r.name = win.name(); } catch (e) {}
  return JSON.stringify(r);
}
]==]

local P = { name = "Safari", bundleID = BUNDLE_ID }

function P.available()
  local path = hs.application.pathForBundleID(BUNDLE_ID)
  return path ~= nil and path ~= ""
end

function P.running()
  return apps.isRunning(BUNDLE_ID)
end

function P.listTabs(cb)
  jxa.run(LIST, { BUNDLE_ID }, cb)
end

function P.activate(tab, cb)
  local argv = { BUNDLE_ID, tostring(tab.windowID), tostring(tab.tabIndex), tostring(tab.url or "") }
  jxa.run(ACTIVATE, argv, function(data, err)
    if err then cb(false, err) return end
    cb(type(data) == "table" and data.ok == true, nil, type(data) == "table" and data or nil)
  end)
end

return P

File: dotfiles/hammerspoon/.hammerspoon/Spoons/BrowserTabs.spoon/test/browsers/safari.js
// The Safari adapter for the harness, the same shape as the Chromium one and separate for the
// same reason the providers are separate. Safari has its own dictionary. A tab's title is name, a
// window names its selected tab through an object rather than an index, and a minimized window is
// miniaturized here.
//
// This one carries an extra field the Chromium adapter has no use for. Every window reports
// whether it has a document, because Safari keeps at least one window that has none, is not on
// screen, and is invisible to the accessibility layer. The spoon must drop that window from its
// listing and this adapter must not, since a harness that hid the phantom could never prove the
// spoon hides it.
//
// Called as osascript -l JavaScript safari.js <bundleID> <op> <args...> answering one JSON object.

function run(argv) {
  const bundleID = argv[0];
  const op = argv[1];
  const args = argv.slice(2);

  const app = Application(bundleID);
  if (!app.running()) return JSON.stringify({ ok: false, err: "not running" });

  function windowById(id) {
    const want = String(id);
    const wins = app.windows;
    for (let i = 0; i < wins.length; i++) {
      try { if (String(wins[i].id()) === want) return wins[i]; } catch (e) {}
    }
    return null;
  }

  function hasDocument(win) {
    try { return !!win.document(); } catch (e) { return false; }
  }

  function describe(win, index) {
    const out = { index: index, id: null, bounds: null, minimized: null, activeTabIndex: -1,
                  document: hasDocument(win), tabs: [] };
    try { out.id = String(win.id()); } catch (e) {}
    try { out.minimized = win.miniaturized(); } catch (e) {}
    try { out.activeTabIndex = win.currentTab().index(); } catch (e) {}
    try {
      const b = win.bounds();
      out.bounds = { x: b.x, y: b.y, w: b.width, h: b.height };
    } catch (e) {}
    let titles = [], urls = [];
    try { titles = win.tabs.name(); urls = win.tabs.url(); } catch (e) { return out; }
    for (let t = 0; t < titles.length; t++) {
      out.tabs.push({ index: t + 1, title: titles[t], url: urls[t] });
    }
    return out;
  }

  const ops = {};

  // Every window, phantom included, since proving the phantom is filtered means seeing it here.
  ops.list = function () {
    const out = [];
    const wins = app.windows;
    for (let i = 0; i < wins.length; i++) out.push(describe(wins[i], i + 1));
    return { ok: true, windows: out };
  };

  // The front window as Safari itself orders them. The phantom has never been observed in front,
  // but it is reported rather than skipped so a round can say so if it ever is.
  ops.front = function () {
    const wins = app.windows;
    if (wins.length === 0) return { ok: true, window: null };
    return { ok: true, window: describe(wins[0], 1) };
  };

  ops.select = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    win.currentTab = win.tabs[parseInt(args[1], 10) - 1];
    return { ok: true };
  };

  ops.raise = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    win.index = 1;
    return { ok: true };
  };

  ops.back = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    win.index = app.windows.length;
    return { ok: true };
  };

  ops.minimize = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    win.miniaturized = args[1] === "1";
    return { ok: true };
  };

  ops.open = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    win.tabs.push(app.Tab({ url: args[1] }));
    return { ok: true, index: win.tabs.length };
  };

  ops.insert = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    const at = parseInt(args[1], 10);
    win.tabs.insert(app.Tab({ url: args[2] }), { at: win.tabs[at - 1] });
    return { ok: true, index: at };
  };

  // A new window, identified by which id is new rather than by taking whatever is in front. Writing
  // to the front window right after asking for a new one navigates whichever window the browser
  // still had there, and this harness once did that to a real window of the person's and then
  // closed it as one of its own.
  ops.openwin = function () {
    const before = {};
    const wins = app.windows;
    for (let i = 0; i < wins.length; i++) {
      try { before[String(wins[i].id())] = true; } catch (e) {}
    }

    // The address is given to the document as it is made. Setting it afterwards through the new
    // window silently does nothing here, which left the window sitting on whatever page Safari
    // opens a new window with.
    if (args[0]) { app.Document({ url: args[0] }).make(); } else { app.Document().make(); }

    // Polled rather than waited on a fixed delay, since every read here is an Apple Event round
    // trip and that is the wait.
    let id = null;
    for (let attempt = 0; attempt < 30 && !id; attempt++) {
      const now = app.windows;
      for (let i = 0; i < now.length; i++) {
        let wid = null;
        try { wid = String(now[i].id()); } catch (e) { continue; }
        if (!before[wid]) { id = wid; break; }
      }
    }
    if (!id) return { ok: false, err: "no new window appeared" };

    return { ok: true, id: id };
  };

  // Closing a tab requires proof that it is one of ours. The caller passes a fragment that must
  // appear in the tab's address, and a tab that does not carry it is left alone.
  //
  // This is here rather than in the caller because a caller got it wrong and it cost somebody a
  // window. A disturbance closed a tab by position, on the assumption that the position still held
  // what it had put there, and when the insert before it had quietly failed the position held a
  // real page instead. Closing it emptied the window and the browser closed the window with it, so
  // a guard that only refused to close windows was never going to be enough.
  ops.close = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    const tab = win.tabs[parseInt(args[1], 10) - 1];
    if (!tab) return { ok: false, err: "no such tab" };
    const must = args[2] || "";
    let url = "";
    try { url = String(tab.url() || ""); } catch (e) {}
    if (!must || url.indexOf(must) < 0) {
      return { ok: false, err: "refusing to close a tab this suite did not open, " + url };
    }
    tab.close();
    return { ok: true };
  };

  ops.closewin = function () {
    const win = windowById(args[0]);
    if (!win) return { ok: false, err: "no such window" };
    win.close();
    return { ok: true };
  };

  // Safari has no scriptable way to move a tab between windows, so this says so rather than
  // pretending. The runner reports the case as not covered instead of quietly passing it.
  ops.move = function () {
    return { ok: false, err: "safari cannot move a tab between windows by script" };
  };

  ops.activate = function () {
    app.activate();
    return { ok: true };
  };

  const fn = ops[op];
  if (!fn) return JSON.stringify({ ok: false, err: "unknown op " + op });
  try {
    return JSON.stringify(fn());
  } catch (e) {
    return JSON.stringify({ ok: false, err: String(e) });
  }
}
